doc/cli.1
.\" Process this file with
.\" groff -man -Tascii cli.1
.\"
.TH CLI 1 "December 2009" "CLI 1.1.0"
.SH NAME
cli \- command line interface compiler for C++
.\"
.\"
.\"
.\"--------------------------------------------------------------------
.SH SYNOPSIS
.\"--------------------------------------------------------------------
.B cli
.B [
.I options
.B ]
.I file
.\"
.\"
.\"
.\"--------------------------------------------------------------------
.SH DESCRIPTION
.\"--------------------------------------------------------------------
.B cli
generates C++ implementation and documentation in various formats for a
command line interface defined in the CLI language. For an input file in
the form
.B name.cli
the following is generated. By default or if the
.B --generate-cxx
option is specified, the following C++ files are generated:
.B name.hxx
(header file),
.B name.ixx
(inline file, generated unless the
.B --suppress-inline
option is specified), and
.B name.cxx (source file).
If the
.B --generate-html
option is specified, then the
.B name.html
HTML documentation file is generated. If the
.B --generate-man
option is specified, then the
.B name.1
man page file is generated. When
.B --generate-html
or
.B --generate-man
is specified, the
.B --stdout
option can be used to redirect the output to STDOUT instead of a file.
.\"
.\"
.\"
.\"--------------------------------------------------------------------
.SH OPTIONS
.\"--------------------------------------------------------------------
.\"
.\" The following documentation was generated by CLI, a command
.\" line interface compiler for C++.
.\"
.IP "\fB--help\fP"
Print usage information and exit\.

.IP "\fB--version\fP"
Print version and exit\.

.IP "\fB--include-path\fP|\fB-I\fP \fIdir\fP"
Search \fIdir\fP for bracket-included (\fB<>\fP) options files\.

.IP "\fB--output-dir\fP|\fB-o\fP \fIdir\fP"
Write the generated files to \fIdir\fP instead of the current directory\.

.IP "\fB--generate-modifier\fP"
Generate option value modifiers in addition to accessors\.

.IP "\fB--generate-specifier\fP"
Generate functions for determining whether the option was specified on the
command line\.

.IP "\fB--generate-description\fP"
Generate the option description list that can be examined at runtime\.

.IP "\fB--generate-file-scanner\fP"
Generate the argv_file_scanner implementation\. This scanner is capable of
reading command line arguments from the argv array as well as files
specified with command line options\.

.IP "\fB--suppress-inline\fP"
Generate all functions non-inline\. By default simple functions are made
inline\. This option suppresses creation of the inline file\.

.IP "\fB--ostream-type\fP \fItype\fP"
Output stream type instead of the default \fBstd::ostream\fP that should be
used to print usage and exception information\.

.IP "\fB--suppress-undocumented\fP"
Suppress the generation of documentation entries for undocumented options\.

.IP "\fB--suppress-usage\fP"
Suppress the generation of the usage printing code\.

.IP "\fB--long-usage\fP"
If no short documentation string is provided, use the complete long
documentation string in usage\. By default, in this situation only the first
sentence from the long string is used\.

.IP "\fB--option-length\fP \fIlen\fP"
Indent option descriptions \fIlen\fP characters when printing usage\. This
is useful when you have multiple options classes, potentially in separate
files, and would like their usage to have the same indentation level\.

.IP "\fB--exclude-base\fP"
Exclude base class information from usage and documentation\.

.IP "\fB--cli-namespace\fP \fIns\fP"
Generate the CLI support types in the \fIns\fP namespace (\fBcli\fP by
default)\. The namespace can be nested, for example \fBdetails::cli\fP\. If
the namespace is empty, then the support types are generated in the global
namespace\.

.IP "\fB--generate-cxx\fP"
Generate C++ code\. If neither \fB--generate-man\fP nor
\fB--generate-html\fP is specified, this mode is assumed by default\.

.IP "\fB--generate-man\fP"
Generate documentation in the man page format\.

.IP "\fB--generate-html\fP"
Generate documentation in the HTML format\.

.IP "\fB--hxx-prologue\fP \fItext\fP"
Insert \fItext\fP at the beginning of the generated C++ header file\.

.IP "\fB--ixx-prologue\fP \fItext\fP"
Insert \fItext\fP at the beginning of the generated C++ inline file\.

.IP "\fB--cxx-prologue\fP \fItext\fP"
Insert \fItext\fP at the beginning of the generated C++ source file\.

.IP "\fB--man-prologue\fP \fItext\fP"
Insert \fItext\fP at the beginning of the generated man page file\.

.IP "\fB--html-prologue\fP \fItext\fP"
Insert \fItext\fP at the beginning of the generated HTML file\.

.IP "\fB--hxx-epilogue\fP \fItext\fP"
Insert \fItext\fP at the end of the generated C++ header file\.

.IP "\fB--ixx-epilogue\fP \fItext\fP"
Insert \fItext\fP at the end of the generated C++ inline file\.

.IP "\fB--cxx-epilogue\fP \fItext\fP"
Insert \fItext\fP at the end of the generated C++ source file\.

.IP "\fB--man-epilogue\fP \fItext\fP"
Insert \fItext\fP at the end of the generated man page text\.

.IP "\fB--html-epilogue\fP \fItext\fP"
Insert \fItext\fP at the end of the generated HTML text\.

.IP "\fB--hxx-prologue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the beginning of the generated C++
header file\.

.IP "\fB--ixx-prologue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the beginning of the generated C++
inline file\.

.IP "\fB--cxx-prologue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the beginning of the generated C++
source file\.

.IP "\fB--man-prologue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the beginning of the generated man page
file\.

.IP "\fB--html-prologue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the beginning of the generated HTML
file\.

.IP "\fB--hxx-epilogue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the end of the generated C++ header
file\.

.IP "\fB--ixx-epilogue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the end of the generated C++ inline
file\.

.IP "\fB--cxx-epilogue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the end of the generated C++ source
file\.

.IP "\fB--man-epilogue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the end of the generated man page file\.

.IP "\fB--html-epilogue-file\fP \fIfile\fP"
Insert the content of \fIfile\fP at the end of the generated HTML file\.

.IP "\fB--class\fP \fIfq-name\fP"
Generate the man page or HTML documentation only for the \fIfq-name\fP
options class\. The \fIfq-name\fP name should be a fully-qualified options
class name, for example, \fBapp::options\fP\. To generate documentation for
multiple classes, repeat this option and the documentation will be produced
in the order specified\. This functionality is useful if you need to
assemble documentation from multiple classes in a specific order or to
insert custom documentation between options belonging to different classes\.

.IP "\fB--stdout\fP"
Write output to STDOUT instead of a file\. This option is not valid when
generating C++ code and is normally used to combine generated documentation
for several option classes in a single file\.

.IP "\fB--hxx-suffix\fP \fIsuffix\fP"
Use \fIsuffix\fP instead of the default \fB\.hxx\fP to construct the name of
the generated header file\.

.IP "\fB--ixx-suffix\fP \fIsuffix\fP"
Use \fIsuffix\fP instead of the default \fB\.ixx\fP to construct the name of
the generated inline file\.

.IP "\fB--cxx-suffix\fP \fIsuffix\fP"
Use \fIsuffix\fP instead of the default \fB\.cxx\fP to construct the name of
the generated source file\.

.IP "\fB--man-suffix\fP \fIsuffix\fP"
Use \fIsuffix\fP instead of the default \fB\.1\fP to construct the name of
the generated man page file\.

.IP "\fB--html-suffix\fP \fIsuffix\fP"
Use \fIsuffix\fP instead of the default \fB\.html\fP to construct the name
of the generated HTML file\.

.IP "\fB--option-prefix\fP \fIprefix\fP"
Use \fIprefix\fP instead of the default \fB-\fP as an option prefix\.
Unknown command line arguments that start with this prefix are treated as
unknown options\. If you set the option prefix to the empty value, then all
the unknown command line arguments will be treated as program arguments\.

.IP "\fB--option-separator\fP \fIsep\fP"
Use \fIsep\fP instead of the default \fB--\fP as an optional separator
between options and arguments\. All the command line arguments that are
parsed after this separator are treated as program arguments\. Set the
option separator to the empty value if you don't want this functionality\.

.IP "\fB--include-with-brackets\fP"
Use angle brackets (<>) instead of quotes ("") in the generated
\fB#include\fP directives\.

.IP "\fB--include-prefix\fP \fIprefix\fP"
Add \fIprefix\fP to the generated \fB#include\fP directive paths\.

.IP "\fB--guard-prefix\fP \fIprefix\fP"
Add \fIprefix\fP to the generated header inclusion guards\. The prefix is
transformed to upper case and characters that are illegal in a preprocessor
macro name are replaced with underscores\.

.IP "\fB--reserved-name\fP \fIname\fP=\fIrep\fP"
Add \fIname\fP with an optional \fIrep\fP replacement to the list of names
that should not be used as identifiers\. If provided, the replacement name
is used instead\. All C++ keywords are already in this list\.

.IP "\fB--options-file\fP \fIfile\fP"
Read additional options from \fIfile\fP with each option appearing on a
separate line optionally followed by space and an option value\. Empty lines
and lines starting with \fB#\fP are ignored\. Option values can be enclosed
in double (\fB"\fP) or single (\fB'\fP) quotes  to preserve leading and
trailing whitespaces as well as to specify empty values\. If the value
itself contains trailing or leading quotes, enclose it with an extra pair of
quotes, for example \fB'"x"'\fP\. Non-leading and non-trailing quotes are
interpreted as being part of the option value\.

The semantics of providing options in a file is equivalent to providing the
same set of options in the same order on the command line at the point where
the \fB--options-file\fP option is specified except that the shell escaping
and quoting is not required\. Repeat this option to specify more than one
options file\.

.\"
.\" DIAGNOSTICS
.\"
.SH DIAGNOSTICS
If the input file is not a valid CLI definition,
.B cli
will issue diagnostic messages to STDERR and exit with non-zero exit code.
.\"
.\" BUGS
.\"
.SH BUGS
Send bug reports to the cli-users@codesynthesis.com mailing list.
.\"
.\" COPYRIGHT
.\"
.SH COPYRIGHT
Copyright (c) 2009-2011 Code Synthesis Tools CC.

Permission is granted to copy, distribute and/or modify this document under
the terms of the MIT License. Copy of this license can be obtained from
http://www.codesynthesis.com/licenses/mit.txt


NEWS
Version 1.2.0

  * New option, --generate-specifier, triggers the generation of functions
    for determining whether the option was specified on the command line.

  * New option, --suppress-undocumented, suppresses the generation of
    documentation entries for undocumented options.

  * New option, --cli-namespace, allows changing of the namespace for the
    generated CLI support types.

  * The argv_file_scanner now supports double and single-quoting option
    values in option files. This is useful to preserve leading and trailing
    whitespaces as well as to specify empty values.

  * The argv_file_scanner now supports multiple file options as well as
    file search callbacks.

Version 1.1.0

  * Support for option documentation. Option documentation is used to print 
    the usage information as well as to generate the program documentation in
    the HTML and man page formats. For details, see Sections 2.5, "Adding
    Documentation" and 3.3, "Option Documentation" in the Getting Started 
    Guide. New CLI compiler command line options related to this feature: 

    --suppress-usage
    --long-usage
    --option-length
    --generate-cxx
    --generate-man
    --generate-html
    --man-prologue
    --man-epilogue
    --html-prologue
    --html-epilogue
    --man-suffix
    --html-suffix
    --class 
    --stdout

    The CLI compiler usage, HTML documentation, and man page are auto-generated
    using this feature.

  * New option, --generate-modifier, triggers the generation of the option
    value modifiers in addition to the accessors.

  * Support for erasing the parsed elements from the argc/argv array. See
    Section 3.1, "Option Class Definition" in the Getting Started Guide for
    more information.

  * New scanner interface. Starting with this version, the option class has 
    a new constructor which accepts an abstract scanner interface. See Section
    3.1, "Option Class Definition" in the Getting Started Guide for more 
    information.

  * New option, --generate-file-scanner, triggers the generation of the 
    argv_file_scanner scanner implementation which provides support for 
    reading command line arguments from the argv array as well as files
    specified with command line options. For more information see Section 
    3.1, "Option Class Definition" in the Getting Started Guide as well as
    the 'file' example.

  * New option, --options-file, allows additional CLI command line options
    to be provided in files (implemented using argv_file_scanner).

Version 1.0.0

  * First public release.


cli/cli.cxx
// file      : cli/cli.cxx
// author    : Boris Kolpackov <boris@codesynthesis.com>
// copyright : Copyright (c) 2009-2011 Code Synthesis Tools CC
// license   : MIT; see accompanying LICENSE file

#include <vector>
#include <string>
#include <memory>   // std::auto_ptr
#include <fstream>
#include <iostream>

#include <cutl/compiler/code-stream.hxx>

#include "options.hxx"
#include "parser.hxx"
#include "generator.hxx"

using namespace std;
using namespace cutl;

int
main (int argc, char* argv[])
{
  ostream& e (cerr);
  const char* file (0);

  try
  {
    cli::argv_file_scanner scan (argc, argv, "--options-file");
    options ops (scan);

    // Handle --version
    //
    if (ops.version ())
    {
      ostream& o (cout);

      o << "CodeSynthesis CLI command line interface compiler 1.1.0" << endl
        << "Copyright (c) 2009-2011 Code Synthesis Tools CC" << endl;

      o << "This is free software; see the source for copying conditions. "
        << "There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS "
        << "FOR A PARTICULAR PURPOSE." << endl;

      return 0;
    }

    // Handle --help
    //
    if (ops.help ())
    {
      ostream& o (cout);

      o << "Usage: " << argv[0] << " [options] file" << endl
        << "Options:" << endl;

      options::print_usage (o);

      return 0;
    }

    if (!scan.more ())
    {
      e << "error: no input file specified" << endl
        << "info: try '" << argv[0] << " --help' for more information" << endl;

      return 1;
    }

    // Extract include search paths.
    //
    parser::paths include_paths;
    for (vector<string>::const_iterator i (ops.include_path ().begin ());
         i != ops.include_path ().end (); ++i)
    {
      // Invalid path exception is handled below.
      //
      include_paths.push_back (semantics::path (*i));
    }

    // Open the input file.
    //
    file = scan.next ();
    semantics::path path (file);

    ifstream ifs (path.string ().c_str ());
    if (!ifs.is_open ())
    {
      e << path << ": error: unable to open in read mode" << endl;
      return 1;
    }

    ifs.exceptions (ifstream::failbit | ifstream::badbit);

    // Parse and generate.
    //
    parser p (include_paths);
    auto_ptr<semantics::cli_unit> unit (p.parse (ifs, path));

    generator g;
    g.generate (ops, *unit, path);
  }
  catch (cli::exception const& ex)
  {
    e << ex << endl;
    return 1;
  }
  catch (semantics::invalid_path const& ex)
  {
    e << "error: '" << ex.path () << "' is not a valid filesystem path"
      << endl;
    return 1;
  }
  catch (std::ios_base::failure const&)
  {
    e << file << ": error: read failure" << endl;
    return 1;
  }
  catch (parser::invalid_input const&)
  {
    // Diagnostics has already been issued by the parser.
    //
    return 1;
  }
  catch (generator::failed const&)
  {
    // Diagnostics has already been issued by the generator.
    //
    return 1;
  }
}
